gradient/src/main.rs

```diff
@@ -0,0 +1,164 @@
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::io::{BufReader, BufWriter};
+
+use anyhow::Result;
+use elevation::GeoTiffElevation;
+use fs_err::File;
+use geo::{Coord, HaversineLength, LineString};
+use geojson::{Feature, FeatureWriter, Geometry};
+use indicatif::{ParallelProgressIterator, ProgressBar, ProgressStyle};
+use osm_reader::{Element, NodeID, WayID};
+use rayon::prelude::*;
+
+/// This takes an osm.pbf file and a GeoTIFF file covering the area. The GeoTIFF must be a digital
+/// elevation model in WGS84 and units of meters. The output is a GeoJSON file with LineStrings for
+/// every way, with a `gradient` property.
+///
+/// Gradient is:
+///
+/// - positive when the edge is uphill from start to end based on the LineString order, and
+///   negative when downhill
+/// - expressed as an integer percent multiplied by 100 -- a value of `305` means a `3.05%` grade.
+fn main() {
+    let args: Vec<String> = std::env::args().collect();
+    if args.len() != 3 {
+        panic!("Call with the input path to an osm.pbf and a GeoTIFF file");
+    }
+
+    run(&args[1], &args[2], "tmp/gradient.geojson").unwrap();
+}
+
+fn run(osm_path: &str, tif_path: &str, output_path: &str) -> Result<()> {
+    let edges = scrape_osm(osm_path)?;
+
+    println!("Calculating gradient");
+    let progress = ProgressBar::new(edges.len() as u64).with_style(ProgressStyle::with_template(
+        "[{elapsed_precise}] [{wide_bar:.cyan/blue}] {human_pos}/{human_len} ({per_sec}, {eta})").unwrap());
+    // Calculate gradient in parallel.
+    let gradients: Vec<(LineString, isize)> = edges
+        .into_par_iter()
+        .progress_with(progress)
+        .map(|(_, linestring)| {
+            // The geotiff reader can only be used from one thread, so open the file about once per
+            // thread (based on how rayon decides to split the work).
+            thread_local!(static ELEVATION: RefCell<Option<GeoTiffElevation<BufReader<File>>>> = RefCell::new(None));
+            ELEVATION.with(|elevation_cell| {
+                let mut elevation = elevation_cell.borrow_mut();
+                if elevation.is_none() {
+                    *elevation = Some(GeoTiffElevation::new(BufReader::new(File::open(tif_path).unwrap())));
+                }
+
+                let gradient = get_steepness(&linestring, elevation.as_mut().unwrap());
```

Comment:
I don't really understand why we want elevation to be mutable, if that is what this means.

Reply:
The method we're calling on this later takes `&mut self`: https://github.com/dabreegster/elevation/blob/e8d0fd4e11774b6204474be2ecb5ad7d05636f2e/src/lib.rs#L26

That's because the underlying call in https://github.com/pka/georaster/ takes `&mut`. In other words, to look up one elevation value with all of this is a mutable operation. That's kind of surprising, but the underlying implementation of the TIFF reader has a stateful implementation that seeks through the file. All of this is annoying, because then we can't have one tiff reader that's shared by many threads. But it's cheap to open a new reader, so it's fine